File: app/src/main/java/com/purpletealabs/bitcoinapp/adapters/CountryListAdapter.java
package com.purpletealabs.bitcoinapp.adapters;

import android.databinding.ObservableArrayList;
import android.databinding.ObservableList;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.purpletealabs.bitcoinapp.databinding.RowCountryBinding;
import com.purpletealabs.bitcoinapp.dtos.Currency;
import com.purpletealabs.bitcoinapp.viewmodels.CountryViewModel;

public class CountryListAdapter extends RecyclerView.Adapter<CountryListAdapter.ViewHolder> {
    private final ObservableArrayList<Currency> countries;
    private final ClickListener listener;

    public CountryListAdapter(ObservableArrayList<Currency> countries, ClickListener listener) {
        this.countries = countries;
        this.listener = listener;
        this.countries.addOnListChangedCallback(new ObservableList.OnListChangedCallback<ObservableList<Currency>>() {
            @Override
            public void onChanged(ObservableList<Currency> sender) {
                notifyDataSetChanged();
            }

            @Override
            public void onItemRangeChanged(ObservableList<Currency> sender, int positionStart, int itemCount) {
                notifyItemRangeChanged(positionStart, itemCount);
            }

            @Override
            public void onItemRangeInserted(ObservableList<Currency> sender, int positionStart, int itemCount) {
                notifyItemRangeInserted(positionStart, itemCount);
            }

            @Override
            public void onItemRangeMoved(ObservableList<Currency> sender, int fromPosition, int toPosition, int itemCount) {
                notifyDataSetChanged();
            }

            @Override
            public void onItemRangeRemoved(ObservableList<Currency> sender, int positionStart, int itemCount) {
                notifyItemRangeRemoved(positionStart, itemCount);
            }
        });
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(RowCountryBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        final Currency currency = countries.get(position);
        holder.binding.setVm(new CountryViewModel(currency.getCurrency(), currency.getCountry()));
    }

    @Override
    public int getItemCount() {
        return countries.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private RowCountryBinding binding;

        ViewHolder(RowCountryBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            this.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onCountrySelected(countries.get(getAdapterPosition()));
                }
            });
        }
    }

    public interface ClickListener {
        void onCountrySelected(Currency currency);
    }
}

File: app/src/main/java/com/purpletealabs/bitcoinapp/adapters/CountryListAdapter.kt
package com.purpletealabs.bitcoinapp.adapters

import android.content.Context
import android.databinding.ObservableArrayList
import android.databinding.ObservableList
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.purpletealabs.bitcoinapp.databinding.RowCountryBinding
import com.purpletealabs.bitcoinapp.dtos.Currency
import com.purpletealabs.bitcoinapp.viewmodels.CountryViewModel

class CountryListAdapter(internal var context: Context, private val countries: ObservableArrayList<Currency>, private val listener: ClickListener) : RecyclerView.Adapter<CountryListAdapter.ViewHolder>() {
    init {
        this.countries.addOnListChangedCallback(object : ObservableList.OnListChangedCallback<ObservableList<Currency>>() {
            override fun onChanged(sender: ObservableList<Currency>) {
                notifyDataSetChanged()
            }

            override fun onItemRangeChanged(sender: ObservableList<Currency>, positionStart: Int, itemCount: Int) {
                notifyItemRangeChanged(positionStart, itemCount)
            }

            override fun onItemRangeInserted(sender: ObservableList<Currency>, positionStart: Int, itemCount: Int) {
                notifyItemRangeInserted(positionStart, itemCount)
            }

            override fun onItemRangeMoved(sender: ObservableList<Currency>, fromPosition: Int, toPosition: Int, itemCount: Int) {
                notifyDataSetChanged()
            }

            override fun onItemRangeRemoved(sender: ObservableList<Currency>, positionStart: Int, itemCount: Int) {
                notifyItemRangeRemoved(positionStart, itemCount)
            }
        })
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(RowCountryBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val currency = countries[position]
        val filename = "_" + currency.currency?.toLowerCase()
        val resID = context.resources.getIdentifier(filename, "drawable", context.packageName)
        holder.binding.vm = CountryViewModel(currency.currency, currency.country, resID)
        holder.binding.executePendingBindings()
    }

    override fun getItemCount(): Int {
        return countries.size
    }

     inner class ViewHolder(var binding: RowCountryBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
            this.itemView.setOnClickListener { listener.onCountrySelected(countries[adapterPosition]) }
        }
    }

    interface ClickListener {
        fun onCountrySelected(currency: Currency)
    }
}